### PR TITLE
[Enhancement] AssessmentDueDate for programming exercises

### PR DESCRIPTION
**Pull request by fde312**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

The assessment due date was not fully implemented for programming exercises. Furthermore, this solves also an issue which we encountered during testing of #2131. Student could see in the response of the REST call the result of his submission even though he should not be able to as the assessment due date was not passed.

### Description

- Added check for assessmentDueDate in `findResultsFilteredForStudents` for the exercise `/details` REST call
- Implemented override for `findResultsFilteredForStudents` for programming exercises `/for-dashboard` REST call
- Implemented override for `findLatestSubmissionWithRatedResultWithCompletionDate` 
- Adapted database query to check for assessmentDueDate for the Student CodeEditor
- Added label to indicate that the assessment due date has passed and that the results will be published immediately (when assessing a exercise for the first time after the assessment due date)
- Remove result button on scores page when assessment due date is in the past (programming exercises)

### Steps for Testing

**Alert and functionality in assessment view when assessment due date is in the past**
1. Log in to Artemis
2. Navigate to Course Administration
3. Create a exercise (Programming, File Upload, Modeling and Text)
4. Participate on exercise
5. Set Assessment Due Date
6. Check if alert is there which indicates that first assessment of submission will be published immediately when assessment due date is in the past
7. Check (as tutor) that after you clicked on submit the override button is not displayed anymore (as instructor you can override as usual)
8. Go to the scores page of the programming exercise, there the result button should not be displayed anymore when the assessment due date is in the past

**Publishing of results before assessment due date and after assessment due date (student)**
1. Log in to Artemis as tutor/instructor
2. Navigate to Course Administration
3. Remove assessment due date from created exercises and assess the submissions
4. Login as student, but keep logged in as well as tutor/instructor
5. Go to course overview -> select your course -> check if you can see the assessed result (course-exercise view + course exercise detail view + student submission view (e.g. Student Code Editor)
6. Check that the assessed result is displayed
7. Now set a assessment due date in the future
8. Check that for Modeling, FileUpload and Text no result is displayed at all. For Programming you should see the automatic result
9. Set the assessment due date now in the past
10. Check that for all exercises the latest manual result is displayed.
11. Check also network tab for the three views, that no result is send out before the assessment due date :) 

### Test Coverage

* ExerciseService.java: 85%
* programming-exercise.component.ts 95%

### Screenshots

**Alert and functionality in assessment view when assessment due date is in the past (tutor/instructor)**

Alert that indicates that the first assessment of a submission will be published immediately:
image

No override button when assessment due date is in the past:
image

Result button on scores pages before assessment due date:
image

..and after assessment due date:
image

**Publishing of results before assessment due date and after assessment due date (student)**

Before assessment due date - Course exercise view:
image

Before assessment due date - Course exercise detail view:
image

Before assessment due date - Student submission view:
image

After assessment due date - Course exercise view:
image
After assessment due date - Course exercise detail view:
image

After assessment due date - Student submission view:
image

### CODE REVIEW

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseParticipationIntegrationTest.java [hunk 1]
```diff
@@ -69,28 +74,81 @@ public void tearDown() {
         database.resetDatabase();
     }
 
-    @Test
+    private static Stream<Arguments> setArgumentsForGetParticipationWithLatestResult() {
```
Comment (MaisiKoleni) [inline src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseParticipationIntegrationTest.java:77]: See reason given by Codacy.
```suggestion
    private static Stream<Arguments> argumentsForGetParticipationWithLatestResult() {
```

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseParticipationIntegrationTest.java [hunk 2]
```diff
@@ -69,28 +74,81 @@ public void tearDown() {
...
+                Arguments.of(AssessmentType.MANUAL, null, null, false), Arguments.of(AssessmentType.MANUAL, null, futureDate, false),
+                Arguments.of(AssessmentType.MANUAL, null, pastDate, false),
+                // Manual result is not returned if completed and assessment due date has not passed
+                Arguments.of(AssessmentType.MANUAL, someDate, futureDate, false),
+                // Manual result is returned if completed and assessmentDue date has passed
+                Arguments.of(AssessmentType.MANUAL, someDate, pastDate, true));
+    }
+
+    @ParameterizedTest
+    @MethodSource("setArgumentsForGetParticipationWithLatestResult")
```
Comment (MaisiKoleni) [inline src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseParticipationIntegrationTest.java:98]:
```suggestion
    @MethodSource("argumentsForGetParticipationWithLatestResult")
```

**Review by MaisiKoleni — COMMENTED** [review]
See inline comments.

File: src/main/java/de/tum/in/www1/artemis/domain/Exercise.java [hunk 3]
```diff
@@ -591,7 +591,8 @@ else if (resultDate1.isAfter(resultDate2)) {
      * @return all results of given participation, or null, if none exist
      */
     public Set<Result> findResultsFilteredForStudents(Participation participation) {
-        return participation.getResults().stream().filter(result -> result.getCompletionDate() != null).collect(Collectors.toSet());
+        boolean isAssessmentOver = getAssessmentDueDate() == null || getAssessmentDueDate().isBefore(ZonedDateTime.now());
+        return participation.getResults().stream().filter(result -> result.getCompletionDate() != null && isAssessmentOver).collect(Collectors.toSet());
```
Comment (sleiss) [inline src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:595]:
```suggestion
        if (!isAssessmentOver) {
            return Set.of();
        }
        return participation.getResults().stream().filter(result -> result.getCompletionDate() != null && isAssessmentOver).collect(Collectors.toSet());
```

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseParticipationIntegrationTest.java [hunk 4]
```diff
@@ -69,28 +74,81 @@ public void tearDown() {
...
+        programmingExerciseRepository.save(programmingExercise);
+        addStudentParticipationWithResult(assessmentType, completionDate);
+        StudentParticipation participation = studentParticipationRepository.findAll().get(0);
+        var expectedStatus = expectLastCreatedResult ? HttpStatus.OK : HttpStatus.NOT_FOUND;
+        request.get(participationsBaseUrl + participation.getId() + "/student-participation-with-latest-result-and-feedbacks", expectedStatus,
                 ProgrammingExerciseStudentParticipation.class);
     }
 
+    @ParameterizedTest
+    @MethodSource("setArgumentsForGetParticipationWithLatestResult")
```
Comment (MaisiKoleni) [inline src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseParticipationIntegrationTest.java:112]:
```suggestion
    @MethodSource("argumentsForGetParticipationWithLatestResult")
```

**Review by sleiss — APPROVED** [review]
Tested it on TS1, worked fine for programming exercises.

**Review by kloessst — APPROVED** [review]
Worked fine in the testing session. Code also looks good! Well done.

**Review by MaisiKoleni — APPROVED** [review]
Works well (tested on TS1, also checked traffic / sent JSON) and code looks good to me now.

**Review by krusche — APPROVED** [review]

**Pull request merged**